File: apps/firehub-api/src/main/java/com/smartfirehub/pipeline/service/executor/ApiCallExecutor.java
package com.smartfirehub.pipeline.service.executor;

import com.smartfirehub.apiconnection.dto.ApiConnectionResponse;
import com.smartfirehub.apiconnection.service.UrlUtils;
import com.smartfirehub.dataset.service.DataTableRowService;
import com.smartfirehub.dataset.service.DataTableService;
import com.smartfirehub.global.exception.ExternalServiceException;
import java.net.URI;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.core.io.buffer.DataBufferLimitException;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.client.reactive.ReactorClientHttpConnector;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.ExchangeStrategies;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.UriComponentsBuilder;
import reactor.netty.http.client.HttpClient;

/**
 * @deprecated Use {@link ExecutorClient#executeApiCall} instead. Retained for
 *     ApiCallPreviewService.
 */
@Deprecated
@Slf4j
@Service
@RequiredArgsConstructor
public class ApiCallExecutor {

  private static final int DEFAULT_TIMEOUT_MS = 30_000;
  private static final int DEFAULT_MAX_DURATION_MS = 3_600_000;
  private static final int DEFAULT_MAX_RETRIES = 3;
  private static final int DEFAULT_INITIAL_BACKOFF = 1_000;
  private static final int DEFAULT_MAX_BACKOFF = 30_000;
  private static final int DEFAULT_PAGE_SIZE = 100;
  private static final int DEFAULT_MAX_RESPONSE_SIZE_MB = 10;
  private static final int MAX_REDIRECTS = 5;

  private final SsrfProtectionService ssrfProtectionService;
  private final JsonResponseParser jsonResponseParser;
  private final OffsetPaginationHandler offsetPaginationHandler;
  private final DataTableService dataTableService;
  private final DataTableRowService dataTableRowService;
  private final WebClient.Builder webClientBuilder;

  // -------------------------------------------------------------------------
  // Result type
  // -------------------------------------------------------------------------

  public record ApiCallResult(int totalRows, String log) {}

  // -------------------------------------------------------------------------
  // Public entry point
  // -------------------------------------------------------------------------

  /**
   * Executes an API_CALL pipeline step.
   *
   * @param config parsed step configuration
   * @param outputTableName target data-schema table name
   * @param decryptedAuthConfig auth key/value pairs (already decrypted), may be null
   * @param loadStrategy "APPEND" or "REPLACE" (REPLACE truncates first)
   * @param conn API 연결 엔티티 (apiConnectionId 설정 시 baseUrl 제공용). null이면 customUrl 또는 레거시 url 사용.
   * @return result with total rows inserted and an execution log string
   */
  public ApiCallResult execute(
      ApiCallConfig config,
      String outputTableName,
      Map<String, String> decryptedAuthConfig,
      String loadStrategy,
      Map<String, String> columnTypeMap,
      ApiConnectionResponse conn) {

    // 1. URL 결정: apiConnectionId → baseUrl+path, customUrl, 또는 레거시 url 순으로 해석
    String resolvedUrl = resolveTargetUrl(config, conn);

    // 2. SSRF guard on the initial URL
    ssrfProtectionService.validateUrl(resolvedUrl);

    // 2. Apply load strategy — for REPLACE create a temp table and insert there;
    //    swap into place only after all pages succeed.
    boolean isReplace = "REPLACE".equalsIgnoreCase(loadStrategy);
    boolean swapped = false;

    if (isReplace) {
      dataTableService.createTempTable(outputTableName);
    }

    // The actual insert target: temp table for REPLACE, original for APPEND
    String insertTarget = isReplace ? outputTableName + "_tmp" : outputTableName;

    long startTime = System.currentTimeMillis();
    int maxDurationMs =
        config.maxDurationMs() != null ? config.maxDurationMs() : DEFAULT_MAX_DURATION_MS;
    int timeoutMs = config.timeoutMs() != null ? config.timeoutMs() : DEFAULT_TIMEOUT_MS;

    int totalRows = 0;
    int totalPages = 0;
    StringBuilder executionLog = new StringBuilder();

    try {
      // 3. Determine pagination mode
      String paginationType = (config.pagination() != null) ? config.pagination().type() : "NONE";

      if ("OFFSET".equalsIgnoreCase(paginationType)) {
        // Offset pagination loop
        ApiCallConfig.PaginationConfig pag = config.pagination();
        int pageSize = (pag.pageSize() != null) ? pag.pageSize() : DEFAULT_PAGE_SIZE;
        int offset = 0;
        Integer totalCount = null;

        while (true) {
          // Build pagination params
          Map<String, String> pagParams =
              offsetPaginationHandler.buildPaginationParams(
                  pag.offsetParam(), pag.limitParam(), offset, pageSize);

          // Fetch page
          String responseBody =
              executeRequest(config, resolvedUrl, decryptedAuthConfig, pagParams, timeoutMs);

          // Parse data
          List<Map<String, Object>> rows =
              jsonResponseParser.parseAndMap(
                  responseBody, config.dataPath(), config.fieldMappings(), columnTypeMap);

          // Extract totalCount from response if configured (only needed on first page)
          if (totalCount == null && pag.totalPath() != null && !pag.totalPath().isBlank()) {
            try {
              Object raw = jsonResponseParser.readPath(responseBody, pag.totalPath(), Object.class);
              if (raw instanceof Number n) {
                totalCount = n.intValue();
              }
            } catch (ApiCallException e) {
              log.debug(
                  "totalPath '{}' not found in response — using partial-page stop condition",
                  pag.totalPath());
            }
          }

          // Insert batch into the target table (tmp for REPLACE, original for APPEND)
          if (!rows.isEmpty()) {
            List<String> columns = extractColumns(rows);
            dataTableRowService.insertBatch(insertTarget, columns, rows);
            totalRows += rows.size();
          }
          totalPages++;

          log.debug("Fetched page offset={} rows={} totalSoFar={}", offset, rows.size(), totalRows);

          // Check duration
          long elapsed = System.currentTimeMillis() - startTime;
          if (elapsed >= maxDurationMs) {
            executionLog
                .append("[WARN] maxDurationMs (")
                .append(maxDurationMs)
                .append(" ms) exceeded after ")
                .append(totalPages)
                .append(" pages. Partial result.\n");
            break;
          }

          // Check for next page
          if (!offsetPaginationHandler.hasNextPage(offset, pageSize, totalCount, rows.size())) {
            break;
          }

          offset = offsetPaginationHandler.getNextOffset(offset, pageSize);
        }

      } else {
        // Single request (no pagination)
        String responseBody =
            executeRequest(config, resolvedUrl, decryptedAuthConfig, Map.of(), timeoutMs);
        List<Map<String, Object>> rows =
            jsonResponseParser.parseAndMap(
                responseBody, config.dataPath(), config.fieldMappings(), columnTypeMap);

        if (!rows.isEmpty()) {
          List<String> columns = extractColumns(rows);
          dataTableRowService.insertBatch(insertTarget, columns, rows);
          totalRows += rows.size();
        }
        totalPages = 1;
      }

      // 4. All pages succeeded — atomically swap tmp -> original for REPLACE
      if (isReplace) {
        dataTableService.swapTable(outputTableName);
        swapped = true;
      }

    } catch (Exception e) {
      // On failure, clean up the temp table so original data is preserved
      if (isReplace && !swapped) {
        try {
          dataTableService.dropTempTable(outputTableName);
        } catch (Exception dropEx) {
          log.warn("Failed to drop temp table after API call failure: {}", dropEx.getMessage());
        }
      }
      throw e;
    }

    long durationMs = System.currentTimeMillis() - startTime;

    executionLog.insert(
        0,
        "url="
            + resolvedUrl
            + " method="
            + config.method()
            + " pages="
            + totalPages
            + " rows="
            + totalRows
            + " duration="
            + durationMs
            + "ms\n");

    return new ApiCallResult(totalRows, executionLog.toString());
  }

  // -------------------------------------------------------------------------
  // HTTP execution with retry
  // -------------------------------------------------------------------------

  private String executeRequest(
      ApiCallConfig config,
      String resolvedUrl,
      Map<String, String> decryptedAuthConfig,
      Map<String, String> paginationParams,
      int timeoutMs) {

    int maxRetries = DEFAULT_MAX_RETRIES;
    int initialBackoff = DEFAULT_INITIAL_BACKOFF;
    int maxBackoff = DEFAULT_MAX_BACKOFF;

    if (config.retry() != null) {
      if (config.retry().maxRetries() != null) maxRetries = config.retry().maxRetries();
      if (config.retry().initialBackoffMs() != null)
        initialBackoff = config.retry().initialBackoffMs();
      if (config.retry().maxBackoffMs() != null) maxBackoff = config.retry().maxBackoffMs();
    }

    int attempt = 0;
    int backoffMs = initialBackoff;
    Exception lastException = null;

    while (attempt <= maxRetries) {
      try {
        return doHttpRequest(config, resolvedUrl, decryptedAuthConfig, paginationParams, timeoutMs);
      } catch (ApiCallException e) {
        // Non-retryable: propagate immediately
        throw e;
      } catch (Exception e) {
        lastException = e;
        attempt++;
        if (attempt > maxRetries) {
          break;
        }
        int jitter = (int) (Math.random() * 1000);
        int sleepMs = Math.min(backoffMs + jitter, maxBackoff);
        log.warn(
            "Request failed (attempt {}/{}), retrying in {} ms: {}",
            attempt,
            maxRetries,
            sleepMs,
            e.getMessage());
        try {
          Thread.sleep(sleepMs);
        } catch (InterruptedException ie) {
          Thread.currentThread().interrupt();
          throw new ApiCallException("Interrupted during retry backoff", ie);
        }
        backoffMs = Math.min(backoffMs * 2, maxBackoff);
      }
    }

    throw new ApiCallException("API call failed after " + maxRetries + " retries", lastException);
  }

  private String doHttpRequest(
      ApiCallConfig config,
      String resolvedUrl,
      Map<String, String> decryptedAuthConfig,
      Map<String, String> paginationParams,
      int timeoutMs) {

    // Merge static query params + pagination params
    Map<String, String> allQueryParams = new LinkedHashMap<>();
    if (config.queryParams() != null) {
      allQueryParams.putAll(config.queryParams());
    }
    if (paginationParams != null) {
      allQueryParams.putAll(paginationParams);
    }

    // Auth: API_KEY with query param goes into queryParams
    if (decryptedAuthConfig != null) {
      String authType = decryptedAuthConfig.get("authType");
      if ("API_KEY".equals(authType) && "query".equals(decryptedAuthConfig.get("placement"))) {
        String paramName = decryptedAuthConfig.get("paramName");
        String apiKey = decryptedAuthConfig.get("apiKey");
        if (paramName != null && apiKey != null) {
          allQueryParams.put(paramName, apiKey);
        }
      }
    }

    // Build the full URI with query params (resolvedUrl은 이미 SSRF 검증 완료)
    UriComponentsBuilder uriBuilder = UriComponentsBuilder.fromUriString(resolvedUrl);
    allQueryParams.forEach(uriBuilder::queryParam);
    URI requestUri = uriBuilder.build(true).toUri();

    // Fix 1: enforce maxResponseSizeMb via ExchangeStrategies codec limit.
    // If the response body exceeds this limit WebClient throws DataBufferLimitException,
    // which is caught below and wrapped as a non-retryable ApiCallException.
    int maxResponseSizeMb =
        (config.maxResponseSizeMb() != null)
            ? config.maxResponseSizeMb()
            : DEFAULT_MAX_RESPONSE_SIZE_MB;
    ExchangeStrategies exchangeStrategies =
        ExchangeStrategies.builder()
            .codecs(c -> c.defaultCodecs().maxInMemorySize(maxResponseSizeMb * 1024 * 1024))
            .build();

    // Fix 2: disable automatic redirects so every hop is SSRF-validated before following.
    // Without this, a malicious server could 301-redirect to http://169.254.169.254/ and
    // bypass the initial validateUrl() check.
    HttpClient reactorHttpClient = HttpClient.create().followRedirect(false);

    WebClient client =
        webClientBuilder
            .clientConnector(new ReactorClientHttpConnector(reactorHttpClient))
            .exchangeStrategies(exchangeStrategies)
            .build();

    String method = (config.method() != null) ? config.method().toUpperCase() : "GET";

    try {
      return executeWithRedirects(
          client, method, requestUri, config, resolvedUrl, decryptedAuthConfig, timeoutMs, 0);
    } catch (DataBufferLimitException e) {
      throw new ApiCallException(
          "Response exceeded maxResponseSizeMb limit ("
              + maxResponseSizeMb
              + " MB) for: "
              + resolvedUrl,
          e);
    } catch (ApiCallException e) {
      throw e;
    } catch (Exception e) {
      // timeout, connection refused, etc. — retryable
      throw new ExternalServiceException(
          "Request error for " + resolvedUrl + ": " + e.getMessage(), e);
    }
  }

  /**
   * Executes a single HTTP request hop and manually follows redirects after SSRF-validating each
   * Location URL. This prevents open-redirect SSRF attacks where an attacker's server issues a
   * 301/302 to an internal address (e.g. http://169.254.169.254/).
   *
   * <p>Redirect following uses GET semantics (standard browser behaviour for 301/302/303). POST
   * bodies are only sent on the initial request; redirects always use GET.
   *
   * @param redirectCount number of redirects followed so far (guards against infinite loops)
   */
  private String executeWithRedirects(
      WebClient client,
      String method,
      URI requestUri,
      ApiCallConfig config,
      String resolvedUrl,
      Map<String, String> decryptedAuthConfig,
      int timeoutMs,
      int redirectCount) {

    if (redirectCount > MAX_REDIRECTS) {
      throw new ApiCallException(
          "Too many redirects (max " + MAX_REDIRECTS + ") for: " + resolvedUrl);
    }

    WebClient.RequestHeadersSpec<?> requestSpec =
        buildRequestSpec(client, method, requestUri, config, decryptedAuthConfig);

    // Use exchangeToMono so we can inspect both status and headers before consuming the body.
    // We capture the Location header via an AtomicReference because the reactive callback
    // cannot return two values at once.
    AtomicReference<String> locationRef = new AtomicReference<>();

    String body =
        requestSpec
            .exchangeToMono(
                response -> {
                  HttpStatusCode status = response.statusCode();

                  if (status.is3xxRedirection()) {
                    // Capture Location header, drain body without buffering, signal redirect with
                    // null
                    String location = response.headers().asHttpHeaders().getFirst("Location");
                    locationRef.set(location);
                    return response
                        .releaseBody()
                        .thenReturn(""); // sentinel: empty string + locationRef set
                  }

                  if (status.value() == 401) {
                    return response
                        .releaseBody()
                        .then(
                            reactor.core.publisher.Mono.error(
                                new ApiCallException(
                                    "Authentication failed (401): " + requestUri)));
                  }

                  if (status.value() == 400 || status.value() == 403 || status.value() == 404) {
                    return response
                        .releaseBody()
                        .then(
                            reactor.core.publisher.Mono.error(
                                new ApiCallException(
                                    "Non-retryable HTTP error "
                                        + status.value()
                                        + " for: "
                                        + requestUri)));
                  }

                  if (status.is4xxClientError() || status.is5xxServerError()) {
                    // 5xx and other 4xx are retryable — surface as plain RuntimeException
                    return response
                        .bodyToMono(String.class)
                        .flatMap(
                            b ->
                                reactor.core.publisher.Mono.error(
                                    new RuntimeException(
                                        "HTTP "
                                            + status.value()
                                            + " from "
                                            + requestUri
                                            + ": "
                                            + b)));
                  }

                  // 2xx — read the response body (size enforced by ExchangeStrategies codec limit)
                  return response.bodyToMono(String.class);
                })
            .timeout(Duration.ofMillis(timeoutMs))
            .block();

    // Handle redirect: locationRef is set only when a 3xx was received
    String location = locationRef.get();
    if (location != null) {
      log.debug("Redirect {}/{}: {} -> {}", redirectCount + 1, MAX_REDIRECTS, requestUri, location);

      // SSRF-validate the redirect target before following it
      ssrfProtectionService.validateUrl(location);

      URI redirectUri = URI.create(location);
      // Redirects always use GET (standard 301/302/303 behaviour)
      return executeWithRedirects(
          client,
          "GET",
          redirectUri,
          config,
          resolvedUrl,
          decryptedAuthConfig,
          timeoutMs,
          redirectCount + 1);
    }

    return body;
  }

  /**
   * Builds a {@link WebClient.RequestHeadersSpec} for the given method and URI, applying configured
   * headers and auth.
   */
  private WebClient.RequestHeadersSpec<?> buildRequestSpec(
      WebClient client,
      String method,
      URI requestUri,
      ApiCallConfig config,
      Map<String, String> decryptedAuthConfig) {

    WebClient.RequestHeadersSpec<?> requestSpec;

    if ("POST".equals(method)) {
      WebClient.RequestBodySpec bodySpec = client.post().uri(requestUri);
      applyHeaders(bodySpec, config, decryptedAuthConfig);
      if (config.body() != null && !config.body().isBlank()) {
        requestSpec = bodySpec.header("Content-Type", "application/json").bodyValue(config.body());
      } else {
        requestSpec = bodySpec;
      }
    } else {
      // GET (default) and redirect follow-ups
      WebClient.RequestHeadersSpec<?> getSpec = client.get().uri(requestUri);
      applyHeaders(getSpec, config, decryptedAuthConfig);
      requestSpec = getSpec;
    }

    return requestSpec;
  }

  @SuppressWarnings("unchecked")
  private void applyHeaders(
      WebClient.RequestHeadersSpec<?> spec,
      ApiCallConfig config,
      Map<String, String> decryptedAuthConfig) {

    // Custom headers from config
    if (config.headers() != null) {
      config.headers().forEach(spec::header);
    }

    // Auth headers
    if (decryptedAuthConfig != null) {
      String authType = decryptedAuthConfig.get("authType");
      String placement = decryptedAuthConfig.getOrDefault("placement", "header");
      if ("API_KEY".equals(authType) && "header".equals(placement)) {
        String headerName = decryptedAuthConfig.get("headerName");
        String apiKey = decryptedAuthConfig.get("apiKey");
        if (headerName != null && apiKey != null) {
          spec.header(headerName, apiKey);
        }
      } else if ("BEARER".equals(authType)) {
        String token = decryptedAuthConfig.get("token");
        if (token != null) {
          spec.header("Authorization", "Bearer " + token);
        }
      }
    }
  }

  // -------------------------------------------------------------------------
  // Helpers
  // -------------------------------------------------------------------------

  /**
   * 하위 호환용 오버로드 — conn 없이 호출하면 customUrl 또는 레거시 url 필드를 사용한다. ApiCallPreviewService 등 conn 정보가 없는
   * 호출부에서 사용.
   */
  public ApiCallResult execute(
      ApiCallConfig config,
      String outputTableName,
      Map<String, String> decryptedAuthConfig,
      String loadStrategy,
      Map<String, String> columnTypeMap) {
    return execute(config, outputTableName, decryptedAuthConfig, loadStrategy, columnTypeMap, null);
  }

  /**
   * API_CALL 스텝의 호출 대상 URL을 계산한다.
   *
   * <p>우선순위:
   *
   * <ol>
   *   <li>conn != null → conn.baseUrl() + config.path() (path 필수)
   *   <li>config.customUrl() → customUrl 그대로 사용
   *   <li>하위 호환: config.url() (deprecated, 이전 설정 호환)
   * </ol>
   *
   * @param config API_CALL 스텝 설정
   * @param conn API 연결 정보 (null이면 customUrl 또는 레거시 url 사용)
   * @throws ApiCallException 필수 파라미터 누락 시
   */
  private String resolveTargetUrl(ApiCallConfig config, ApiConnectionResponse conn) {
    if (conn != null) {
      // apiConnectionId 기반: connection.baseUrl + path
      String path = config.path();
      if (path == null || path.isBlank()) {
        throw new ApiCallException("API_CALL: apiConnectionId 설정 시 path가 필수입니다");
      }
      return UrlUtils.joinUrl(conn.baseUrl(), path);
    }

    // customUrl 우선 사용 (Phase 9 신규 필드)
    String customUrl = config.customUrl();
    if (customUrl != null && !customUrl.isBlank()) {
      return customUrl;
    }

    // 하위 호환: 레거시 url 필드
    String legacyUrl = config.url();
    if (legacyUrl != null && !legacyUrl.isBlank()) {
      return legacyUrl;
    }

    throw new ApiCallException("API_CALL: apiConnectionId 없이 호출하려면 customUrl(또는 레거시 url)이 필수입니다");
  }

  private List<String> extractColumns(List<Map<String, Object>> rows) {
    if (rows.isEmpty()) return List.of();
    return new ArrayList<>(rows.get(0).keySet());
  }
}

File: apps/firehub-executor/app/services/api_call_executor.py
from __future__ import annotations

import logging
import random
import time
from datetime import date, datetime, timezone
from decimal import Decimal, InvalidOperation
from typing import Any, Dict, List, Optional
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

import httpx
from jsonpath_ng import parse as jsonpath_parse
from psycopg2 import sql as pgsql
from psycopg2.extras import execute_values

from app.schemas.requests import ApiCallExecuteRequest, FieldMapping
from app.schemas.responses import ApiCallExecuteResponse
from app.validators.ssrf_protection import SsrfException, validate_url

logger = logging.getLogger(__name__)

MAX_REDIRECTS = 5
DEFAULT_PAGE_SIZE = 100
MAX_PAGES = 10_000

# Non-retryable HTTP status codes
NON_RETRYABLE_STATUS = {400, 401, 403, 404}


class ApiCallException(Exception):
    """Non-retryable API call error."""
    pass


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------

def execute_api_call(request: ApiCallExecuteRequest, conn) -> ApiCallExecuteResponse:
    start_time = time.monotonic()
    try:
        # 1. SSRF validation on initial URL
        validate_url(request.url)

        total_rows = 0
        total_pages = 0
        log_lines: List[str] = []

        # 2. Pagination
        pagination_type = (request.pagination.type.upper() if request.pagination else "NONE")

        if pagination_type == "OFFSET":
            pag = request.pagination
            page_size = pag.page_size or DEFAULT_PAGE_SIZE
            offset = 0
            total_count: Optional[int] = None

            while total_pages < MAX_PAGES:
                # Build pagination params
                pag_params: Dict[str, str] = {}
                if pag.offset_param:
                    pag_params[pag.offset_param] = str(offset)
                if pag.limit_param:
                    pag_params[pag.limit_param] = str(page_size)

                response_body = _execute_request(request, pag_params)
                rows = _parse_and_map(
                    response_body,
                    request.data_path,
                    request.field_mappings,
                    request.column_type_map,
                )

                # Extract total count (first page only)
                if total_count is None and pag.total_path:
                    raw_total = _read_path_safe(response_body, pag.total_path)
                    if raw_total is not None:
                        try:
                            total_count = int(raw_total)
                        except (TypeError, ValueError):
                            total_count = None

                if rows:
                    _insert_batch(conn, request.output_table, rows)
                    total_rows += len(rows)
                total_pages += 1

                # Duration check
                elapsed_ms = int((time.monotonic() - start_time) * 1000)
                if elapsed_ms >= request.max_duration_ms:
                    log_lines.append(
                        f"[WARN] maxDurationMs ({request.max_duration_ms} ms) exceeded "
                        f"after {total_pages} pages. Partial result."
                    )
                    break

                # Next page check
                if not _has_next_page(offset, page_size, total_count, len(rows)):
                    break
                offset += page_size
        else:
            # Single request (NONE or unknown pagination type)
            response_body = _execute_request(request, {})
            rows = _parse_and_map(
                response_body,
                request.data_path,
                request.field_mappings,
                request.column_type_map,
            )
            if rows:
                _insert_batch(conn, request.output_table, rows)
                total_rows += len(rows)
            total_pages = 1

        elapsed_ms = int((time.monotonic() - start_time) * 1000)
        summary = (
            f"url={request.url} method={request.method} "
            f"pages={total_pages} rows={total_rows} duration={elapsed_ms}ms"
        )
        log_lines.insert(0, summary)

        conn.commit()

        return ApiCallExecuteResponse(
            success=True,
            rows_loaded=total_rows,
            total_pages=total_pages,
            execution_log="\n".join(log_lines),
            execution_time_ms=elapsed_ms,
        )

    except (SsrfException, ApiCallException) as e:
        conn.rollback()
        elapsed_ms = int((time.monotonic() - start_time) * 1000)
        return ApiCallExecuteResponse(
            success=False, error=str(e), execution_time_ms=elapsed_ms
        )
    except Exception as e:
        conn.rollback()
        elapsed_ms = int((time.monotonic() - start_time) * 1000)
        logger.exception("Unexpected error in execute_api_call")
        return ApiCallExecuteResponse(
            success=False, error=str(e), execution_time_ms=elapsed_ms
        )


# ---------------------------------------------------------------------------
# HTTP request with retry + manual redirect + SSRF check
# ---------------------------------------------------------------------------

def _execute_request(request: ApiCallExecuteRequest, extra_params: Dict[str, str]) -> Any:
    """Execute a single HTTP request with retry and manual redirect following."""
    retry_cfg = request.retry
    max_retries = retry_cfg.max_retries if retry_cfg else 3
    initial_backoff_ms = retry_cfg.initial_backoff_ms if retry_cfg else 1000
    max_backoff_ms = retry_cfg.max_backoff_ms if retry_cfg else 30000

    timeout_sec = request.timeout_ms / 1000.0
    max_bytes = request.max_response_size_mb * 1024 * 1024

    attempt = 0
    last_error: Optional[Exception] = None

    while attempt <= max_retries:
        try:
            response_body = _do_request(
                url=request.url,
                method=request.method,
                headers=request.headers or {},
                query_params={**(request.query_params or {}), **extra_params},
                body=request.body,
                auth=request.auth,
                timeout_sec=timeout_sec,
                max_bytes=max_bytes,
            )
            return response_body
        except ApiCallException:
            # Non-retryable — propagate immediately
            raise
        except Exception as e:
            last_error = e
            attempt += 1
            if attempt > max_retries:
                break
            backoff_ms = min(initial_backoff_ms * (2 ** (attempt - 1)), max_backoff_ms)
            # Add jitter: ±25%
            jitter_ms = backoff_ms * 0.25 * (random.random() * 2 - 1)
            sleep_sec = max(0.0, (backoff_ms + jitter_ms) / 1000.0)
            logger.warning(
                "Attempt %d/%d failed (%s). Retrying in %.2fs",
                attempt, max_retries, e, sleep_sec,
            )
            time.sleep(sleep_sec)

    raise ApiCallException(f"All {max_retries + 1} attempts failed. Last error: {last_error}")


def _do_request(
    url: str,
    method: str,
    headers: Dict[str, str],
    query_params: Dict[str, str],
    body: Optional[str],
    auth: Optional[dict],
    timeout_sec: float,
    max_bytes: int,
) -> Any:
    """Perform a single HTTP call with manual redirect following and SSRF checks."""
    # Build auth headers
    req_headers = dict(headers)
    req_params = dict(query_params)
    _apply_auth(auth, req_headers, req_params)

    current_url = url
    redirect_count = 0

    with httpx.Client(follow_redirects=False, timeout=timeout_sec) as client:
        while True:
            resp = client.request(
                method=method,
                url=current_url,
                headers=req_headers,
                params=req_params if redirect_count == 0 else {},
                content=body.encode() if body else None,
            )

            if resp.status_code in (301, 302, 303, 307, 308):
                redirect_count += 1
                if redirect_count > MAX_REDIRECTS:
                    raise ApiCallException(
                        f"Too many redirects (max {MAX_REDIRECTS})"
                    )
                location = resp.headers.get("location")
                if not location:
                    raise ApiCallException("Redirect response missing Location header")
                # SSRF check on redirect target
                validate_url(location)
                current_url = location
                # For 303, switch to GET
                if resp.status_code == 303:
                    method = "GET"
                    body = None
                continue

            # Check for non-retryable client errors
            if resp.status_code in NON_RETRYABLE_STATUS:
                raise ApiCallException(
                    f"Non-retryable HTTP error {resp.status_code}: {resp.text[:200]}"
                )

            # Raise for other HTTP errors (retryable 5xx etc.)
            if resp.status_code >= 400:
                raise Exception(
                    f"HTTP error {resp.status_code}: {resp.text[:200]}"
                )

            # Check response size
            content_length = resp.headers.get("content-length")
            if content_length and int(content_length) > max_bytes:
                raise ApiCallException(
                    f"Response size {content_length} bytes exceeds limit {max_bytes} bytes"
                )

            content = resp.content
            if len(content) > max_bytes:
                raise ApiCallException(
                    f"Response body {len(content)} bytes exceeds limit {max_bytes} bytes"
                )

            return resp.json()


def _apply_auth(
    auth: Optional[dict],
    headers: Dict[str, str],
    params: Dict[str, str],
) -> None:
    """Apply authentication to request headers or query params.

    Supports both Java decryptedAuth format (authType, headerName, apiKey,
    placement, paramName) and generic format (type, key_name, key_value, location).
    """
    if not auth:
        return
    auth_type = (auth.get("authType") or auth.get("type") or "").upper()
    if auth_type == "BEARER":
        token = auth.get("token") or auth.get("value") or ""
        headers["Authorization"] = f"Bearer {token}"
    elif auth_type == "API_KEY":
        placement = (auth.get("placement") or auth.get("location") or "header").upper()
        if placement == "QUERY":
            param_name = auth.get("paramName") or auth.get("key_name") or "api_key"
            api_key = auth.get("apiKey") or auth.get("key_value") or auth.get("value") or ""
            params[param_name] = api_key
        else:
            header_name = auth.get("headerName") or auth.get("key_name") or "x-api-key"
            api_key = auth.get("apiKey") or auth.get("key_value") or auth.get("value") or ""
            headers[header_name] = api_key


# ---------------------------------------------------------------------------
# JSON parsing + field mapping
# ---------------------------------------------------------------------------

def _parse_and_map(
    response_body: Any,
    data_path: str,
    field_mappings: List[FieldMapping],
    column_type_map: Optional[dict],
) -> List[Dict[str, Any]]:
    """Extract items from response using jsonpath, map fields."""
    items = _extract_data(response_body, data_path)
    if not items:
        return []

    rows: List[Dict[str, Any]] = []
    for item in items:
        if not isinstance(item, dict):
            continue
        row: Dict[str, Any] = {}
        for fm in field_mappings:
            raw_value = item.get(fm.source_field)
            converted = _convert_value(raw_value, fm)
            row[fm.target_column] = converted
        rows.append(row)
    return rows


def _extract_data(response_body: Any, data_path: str) -> List[Any]:
    """Use jsonpath_ng to extract list of items at data_path."""
    if not data_path or data_path.strip() in ("$", ""):
        if isinstance(response_body, list):
            return response_body
        return [response_body]

    try:
        expr = jsonpath_parse(data_path)
        matches = expr.find(response_body)
        if not matches:
            return []
        # If the matched value is a list, return it; otherwise wrap
        value = matches[0].value
        if isinstance(value, list):
            return value
        return [value]
    except Exception as e:
        logger.warning("jsonpath parse/find error for path '%s': %s", data_path, e)
        return []


def _read_path_safe(response_body: Any, path: str) -> Any:
    """Read a single value at path, returning None on error."""
    try:
        expr = jsonpath_parse(path)
        matches = expr.find(response_body)
        if matches:
            return matches[0].value
    except Exception:
        pass
    return None


def _convert_value(value: Any, fm: FieldMapping) -> Any:
    """Convert a raw value to the target data type specified in FieldMapping."""
    if value is None:
        return None

    data_type = (fm.data_type or "TEXT").upper()

    try:
        if data_type in ("TEXT", "VARCHAR", "STRING"):
            return str(value)

        if data_type == "INTEGER":
            str_val = str(value)
            if fm.number_format and "comma_separated" in fm.number_format.lower():
                str_val = str_val.replace(",", "")
            return int(str_val.strip())

        if data_type in ("DECIMAL", "NUMERIC", "FLOAT", "DOUBLE"):
            str_val = str(value)
            if fm.number_format and "comma_separated" in fm.number_format.lower():
                str_val = str_val.replace(",", "")
            return Decimal(str_val.strip())

        if data_type == "BOOLEAN":
            if isinstance(value, bool):
                return value
            str_val = str(value).strip().lower()
            return str_val in ("true", "1", "yes")

        if data_type == "DATE":
            if isinstance(value, date):
                return value
            str_val = str(value).strip()
            if fm.date_format:
                return datetime.strptime(str_val, fm.date_format).date()
            # Try ISO format
            return date.fromisoformat(str_val)

        if data_type in ("TIMESTAMP", "DATETIME"):
            if isinstance(value, datetime):
                dt = value
            else:
                str_val = str(value).strip()
                if fm.date_format:
                    dt = datetime.strptime(str_val, fm.date_format)
                else:
                    # Try ISO format
                    dt = datetime.fromisoformat(str_val.replace("Z", "+00:00"))

            # Convert to UTC
            if fm.source_timezone and dt.tzinfo is None:
                try:
                    tz = ZoneInfo(fm.source_timezone)
                    dt = dt.replace(tzinfo=tz)
                except ZoneInfoNotFoundError:
                    pass
            if dt.tzinfo is not None:
                dt = dt.astimezone(timezone.utc).replace(tzinfo=None)
            return dt

        if data_type == "GEOMETRY":
            return str(value)

        # Default: return as-is
        return value

    except (ValueError, InvalidOperation, TypeError) as e:
        logger.warning(
            "Type conversion failed for field '%s' value=%r type=%s: %s",
            fm.source_field, value, data_type, e,
        )
        return None


# ---------------------------------------------------------------------------
# Pagination helper
# ---------------------------------------------------------------------------

def _has_next_page(
    offset: int,
    page_size: int,
    total_count: Optional[int],
    current_page_rows: int,
) -> bool:
    """Determine whether there is a next page."""
    if current_page_rows < page_size:
        # Partial page — no more data
        return False
    if total_count is not None:
        return offset + page_size < total_count
    # No total count: continue as long as we got a full page
    return True


# ---------------------------------------------------------------------------
# DB insert
# ---------------------------------------------------------------------------

def _insert_batch(conn, table_name: str, rows: List[Dict[str, Any]]) -> None:
    """Batch-insert rows into target table using execute_values."""
    if not rows:
        return

    columns = list(rows[0].keys())
    col_identifiers = [pgsql.Identifier(c) for c in columns]

    query = pgsql.SQL("INSERT INTO {table} ({cols}) VALUES %s").format(
        table=pgsql.Identifier(table_name),
        cols=pgsql.SQL(", ").join(col_identifiers),
    )

    values = [tuple(row[c] for c in columns) for row in rows]

    with conn.cursor() as cur:
        execute_values(cur, query.as_string(cur), values)
